sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -777,17 +768,65 @@ object GpuCast extends Arm {
     }
   }
 
+  def castStringToDecimal(
+      input: ColumnView,
+      ansiEnabled: Boolean,
+      dt: DecimalType): ColumnVector = {
+    // 1. Sanitize strings to make sure all are floating points
+    // 2. Identify all fixed point values
+    // 3. Cast String to newDt (newDt = dt.scale + 1). Promote precision if needed. This step is
+    //    required so we can round up if needed in the final step
+    // 4. Now cast to newDt to dt (Decimal to Decimal)
+    def getInterimDecimalPromoteIfNeeded(dt: DecimalType): DecimalType = {
+      if (dt.scale + 1 > dt.precision) {
+        // promote if possible or throw
+        if (dt.precision == Decimal.MAX_LONG_DIGITS) {
+          //We don't support Decimal 128
+          throw new IllegalArgumentException("One or more values exceed the maximum supported " +
+              "Decimal precision while conversion")
+        }
+        DecimalType(dt.precision + 1, dt.scale + 1)
+      }
+      DecimalType(dt.precision, dt.scale + 1)
+    }
+
+    withResource(input.strip()) { trimmed =>
+      withResource(GpuCast.sanitizeStringToFloat(trimmed, ansiEnabled, false)) { sanitized =>
+        val interimSparkDt = getInterimDecimalPromoteIfNeeded(dt)
+        val interimDt = DecimalUtil.createCudfDecimal(interimSparkDt)
+        withResource(Scalar.fromNull(interimDt)) { nulls =>
```

Comment:
nit: can we put the nulls closer to where it is used? There are cases where we will never need to create it, so why bother?